ischroot.1
.\" -*- nroff -*-
.TH ISCHROOT 1 "30 May 2011" "Debian"
.SH NAME
ischroot \- detect if running in a chroot
.SH SYNOPSIS
.B ischroot
[\-\-default\-false] [\-\-default\-true] [\-\-help] [\-\-version]
.SH DESCRIPTION
.PP
.B ischroot
detects if it is currently running in a chroot. The exit status is:
.TP
0
if currently running in a chroot
.TP
1
if currently not running in a chroot
.TP
2
if the detection is not possible (On GNU/Linux this happens if the
script is not run as root).
.SH OPTIONS
.TP
.B "-f, --default-false "
Exit with status 1 if the detection is not possible.
.TP
.B "-t, --default-true "
Exit with status 0 if the detection is not possible.
.TP
.B "--help"
Print a usage message on standard output and exit successfully.
.TP
.B "--version"
Print version information on standard output and exit successfully.
.SH BUGS
On GNU/Linux, chroot detection is not possible when not root. This
works correctly on GNU/Hurd and GNU/kFreeBSD.


debian/changelog
debianutils (4.0.1) unstable; urgency=low

  * tempfile.1: state in the man page that tempfile is
    deprecated.  closes: #627564.
  * ischroot.1: correct root/non-root logic claims.
    closes: #628465.

 -- Clint Adams <clint@debian.org>  Mon, 30 May 2011 10:54:03 -0400

debianutils (4) unstable; urgency=low

  * ischroot utility by Aurelien Jarno to detect chrootedness.
    closes: #627205.

 -- Clint Adams <clint@debian.org>  Wed, 18 May 2011 15:04:15 -0400

debianutils (3.4.6) unstable; urgency=low

  * Add Spanish man page translations from Omar
    Campagne.  closes: #619016.
  * Updated German man page translations from Helge
    Kreutzmann.  closes: #626891.

 -- Clint Adams <clint@debian.org>  Tue, 17 May 2011 23:29:42 -0400

debianutils (3.4.5) unstable; urgency=low

  * Vcs-Git update.
  * add-shell(8): grammar fix from Helge Kreutzmann.  closes: #626409.
  * German man page translations from Helge Kreutzmann.  closes: #626397.
  * Bump to Standards-Version 3.9.2.

 -- Clint Adams <clint@debian.org>  Thu, 12 May 2011 22:45:15 -0400

debianutils (3.4.4) unstable; urgency=low

  * French man page translation update from David Prévot.
    closes: #603908.
  * Patch from Jari Aalto to reorder tempfile options alphabetically.
    closes: #600542.

 -- Clint Adams <clint@debian.org>  Thu, 10 Feb 2011 17:24:34 -0500

debianutils (3.4.3) unstable; urgency=low

  * Patch from Ben Hutchings to have installkernel run hook scripts.
    closes: #607411.

 -- Clint Adams <clint@gnu.org>  Tue, 18 Jan 2011 15:29:35 -0500

debianutils (3.4.2) unstable; urgency=low

  * Add Slovene man page translations from Peter Kragelj.  closes:
    #598531.
  * French man page translation update from Christian Perrier.  closes:
    #597486.
  * run-parts.8: be more precise about ASCII requirement in naming.
    closes: #542029.

 -- Clint Adams <clint@gnu.org>  Sun, 14 Nov 2010 10:47:12 -0500

debianutils (3.4.1) unstable; urgency=low

  * po4a/po/fr.po: French man page translation update from
    Christian Perrier.  closes: #597430.
  * Bump to Standards-Version 3.9.1.

 -- Clint Adams <schizo@debian.org>  Sun, 19 Sep 2010 12:58:38 -0400

debianutils (3.4) unstable; urgency=low

  * Apply patch from Ariel to add strength options for compression and
    reduce xz default compression level.  closes: #587766.
  * Bump Standards-Version to 3.9.0.

 -- Clint Adams <schizo@debian.org>  Thu, 15 Jul 2010 22:02:27 -0400

debianutils (3.3) unstable; urgency=low

  * savelog.8: add SEE ALSO for logrotate(8).  closes: #487171.
  * savelog: add -J option for xz compression.  closes: #587520.

 -- Clint Adams <schizo@debian.org>  Wed, 30 Jun 2010 10:38:10 -0400

debianutils (3.2.3) unstable; urgency=low

  * Switch to 3.0 (native) source format.
  * French man page translation updates from Alexandre Hoïde.
    closes: #579652.
  * Bump to Standards-Version 3.8.4.

 -- Clint Adams <schizo@debian.org>  Thu, 29 Apr 2010 11:12:26 -0400

debianutils (3.2.2) unstable; urgency=low

  * savelog: patch from Patrick Coleman to fix breakage when target file
    does not exist.  closes: #556227.
  * Bump to Standards-Version 3.8.3.

 -- Clint Adams <schizo@debian.org>  Sat, 14 Nov 2009 13:16:01 -0500

debianutils (3.2.1) unstable; urgency=low

  * run-parts.8: conform man page regex with the actual regex used by
    run-parts.  closes: #536395.

 -- Clint Adams <schizo@debian.org>  Sun, 09 Aug 2009 15:18:52 -0400

debianutils (3.2) unstable; urgency=low

  * run-parts.c: print command arguments when --verbose and --arg are
    used.  closes: #535621.
  * Bump to Standards-Version 3.8.2.

 -- Clint Adams <schizo@debian.org>  Sun, 05 Jul 2009 15:03:00 -0400

debianutils (3.1.3) unstable; urgency=low

  * Remove any mention of readlink from debian/copyright.  closes:
    #527138.

 -- Clint Adams <schizo@debian.org>  Tue, 05 May 2009 21:08:31 -0400

debianutils (3.1.2) unstable; urgency=low

  * Patch from Zack Weinberg to fix tempfile -n crash.  closes: #526981.

 -- Clint Adams <schizo@debian.org>  Mon, 04 May 2009 18:56:14 -0400

debianutils (3.1.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * Fix a double free bug in tempfile. Closes: #526606, #526607, #526610

 -- Manoj Srivastava <srivasta@debian.org>  Sat, 02 May 2009 03:45:30 -0500

debianutils (3.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * which.1: clarify that which does not follow symlinks.  closes: #268624.
  * run-parts.c: permission and sanity checks run on link targets, not
    symlinks.  closes: #294789.
  * tempfile.c: tempfile with a suffix no longer needs hardlinks.
    closes: #312270.
  * savelog.8: disambiguate version numbering in manual.  closes: #330289.
  * tempfile.1: update manual page with up-to-date tempnam behaviors.
    closes: #338490.

  [ Clint Adams ]
  * savelog.8, tempfile.1: remove spurious trailing whitespace
  * savelog.8, tempfile.1, which.1: update dates in manual pages.
  * po4a/po/fr.po: tweak French translations enough to reach 80%.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 18:01:28 -0400

debianutils (3.0.2) unstable; urgency=low

  * installkernel: fix symlinking of kernel images.  closes: #526493.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 11:08:13 -0400

debianutils (3.0.1) unstable; urgency=low

  * Remove French sensible-editor man page as well.  closes: #521543.

 -- Clint Adams <schizo@debian.org>  Sun, 29 Mar 2009 17:55:21 -0400

debianutils (3.0) unstable; urgency=low

  * Stop shipping sensible-browser, sensible-editor, sensible-pager, and
    accompanying man pages, and declare a dependency on sensible-utils,
    to which they have moved.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 21:25:08 -0400

debianutils (2.32.3) unstable; urgency=medium

  * installkernel: remove stray right bracket, thanks to Alban Browaeys.
    closes: #521286.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 09:53:46 -0400

debianutils (2.32.2) unstable; urgency=low


ischroot.c
/* ischroot: detect if running in a chroot
 *
 * Debian ischroot program
 * Copyright (C) 2011 Aurelien Jarno <aurel32@debian.org>
 *
 * This is free software; see the GNU General Public License version 2
 * or later for copying conditions.  There is NO warranty.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif /* HAVE_GETOPT_H */

void version()
{
  fprintf(stderr, "Debian ischroot, version " PACKAGE_VERSION
          "Copyright (C) 2011 Aurelien Jarno\n"
          "This is free software; see the GNU General Public License version 2\n"
          "or later for copying conditions.  There is NO warranty.\n");
  exit(0);
}

void usage()
{
  fprintf(stderr, "Usage: ischroot [OPTION]\n"
          "  -f, --default-false return false if detection fails\n"
          "  -t, --default-true  return true if detection fails\n"
          "  -V, --version       output version information and exit.\n"
          "  -h, --help          display this help and exit.\n");
  exit(0);
}

#if defined (__linux__)

/* On Linux we can detect chroots by checking if the 
 * devicenumber/inode pair of / are the same as that of 
 * /sbin/init's. This may fail if not running as root or if
 * /proc is not mounted, in which case 2 is returned.
 */

static int ischroot()
{
  struct stat st1, st2;

  if (stat("/", &st1) || stat("/proc/1/root", &st2))
    return 2;
  else if ((st1.st_dev == st2.st_dev) && (st1.st_ino == st2.st_ino))
    return 1;
  else
    return 0;
}

#elif defined (__FreeBSD_kernel__) || defined (__FreeBSD__)

#include <sys/sysctl.h>
#include <sys/user.h>

/* On FreeBSD we can detect chroot by looking for a specific
 * file descriptor pointing to the location of the chroot. There
 * is not need to be root, so it is unlikely to fail in normal 
 * cases, but return 2 if a memory failure or the like happens. */

static int ischroot()
{
  int mib[4];
  size_t kf_len = 0;
  char *kf_buf, *kf_bufp;

  mib[0] = CTL_KERN;
  mib[1] = KERN_PROC;
  mib[2] = KERN_PROC_FILEDESC;
  mib[3] = getpid ();

  if (sysctl (mib, 4, NULL, &kf_len, NULL, 0) != 0)
    return 2;

  kf_buf = kf_bufp = malloc (kf_len);
  if (kf_buf == NULL)
    return 2;

  if (sysctl (mib, 4, kf_buf, &kf_len, NULL, 0) != 0)
  {
    free(kf_buf);
    return 2;
  }

  while (kf_bufp < kf_buf + kf_len)
  {   
    struct kinfo_file *kf = (struct kinfo_file *) (uintptr_t) kf_bufp;

    if (kf->kf_fd == KF_FD_TYPE_JAIL)
    {
      free(kf_buf);
      return 0;
    }
    kf_bufp += kf->kf_structsize;
  }   

  free(kf_buf);
  return 1;
}

#elif defined (__GNU__)

/* On Hurd we can detect chroot by looking at the device number
 * containing /. The device number of the first mounted filesystem
 * equals 3, and due to bug http://savannah.gnu.org/bugs/?23213
 * chroots have to be created on a different filesystem. Return 2
 * if it is not possible to probe this device. */

static int ischroot()
{
  struct stat st;

  if (stat("/", &st))
    return 2;
  else if (st.st_dev == 3)
    return 1;
  else
    return 0;
}

#else

static int ischroot()
{
  return 2;
}

#warning unknown system, chroot detection will always fail

#endif

/* Process options */
int main(int argc, char *argv[])
{
  int default_false = 0;
  int default_true = 0;
  int exit_status;

  for (;;) {
    int c;
    int option_index = 0;

    static struct option long_options[] = {
      {"default-false", 0, 0, 'f'},
      {"default-true", 0, 0, 't'},
      {"help", 0, 0, 'h'},
      {"version", 0, 0, 'V'},
      {0, 0, 0, 0}
    };
    c = getopt_long(argc, argv, "fthV", long_options, &option_index);
    if (c == EOF)
      break;
    switch (c) {
    case 'f':
      default_false = 1;
      break;
    case 't':
      default_true = 1;
      break;
    case 'h':
      usage();
      break;
    case 'V':
      version();
      break;
    default:
      fprintf(stderr, "Try `ischroot --help' for more information.\n");
      exit(1);
    }
  }

  if (default_false && default_true) {
    fprintf(stderr, "Can't default to both true and false!\n");
    fprintf(stderr, "Try `ischroot --help' for more information.\n");
    exit(1);
  }

  exit_status = ischroot();

  if (exit_status == 2) {
    if (default_true)
      exit_status = 0;
    if (default_false)
      exit_status = 1;
  }

  return exit_status;
}
